corgidrp/calibrate_kgain.py

```diff
@@ -0,0 +1,924 @@
+    nan_avg = np.isnan(logged_averages)
+    
+    # Filter out NaN's from the data
+    logged_deviations = logged_deviations[~nan_dev & ~nan_avg]
+    logged_averages = logged_averages[~nan_dev & ~nan_avg]
+    
+    # Calculate the error factor
+    logged_deviations_error_factor = compiled_binned_deviations_error / compiled_binned_shot_deviations
+    logged_deviations_error_factor = logged_deviations_error_factor[lower_bound:upper_bound]
+    logged_deviations_error_factor = logged_deviations_error_factor[~nan_dev & ~nan_avg]
+    
+    # Calculate the logged deviations error
+    logged_deviations_error = logged_deviations_error_factor * logged_deviations
+    
+    # Create a mask for non-NaN values
+    non_nan_mask = ~np.isnan(logged_deviations_error)
+
+    # Use the mask to filter out NaN values
+    y_err = logged_deviations_error[non_nan_mask]
+    x_vals = logged_averages[non_nan_mask]
+    y_vals = logged_deviations[non_nan_mask]
+    
+    # make array of kgain values from variance and mean values
+    # equation: k_gain = mean / signal variance
+    kgain_arr = 10**(x_vals)/(10**y_vals)**2
+    
+    if make_plot:
+        fname = 'kgain_histogram'
+        plt.figure()
+        # 'auto' lets matplotlib decide the number of bins
+        plt.hist(kgain_arr, bins='auto', log=True)
+        plt.title('Histogram of kgain values')
+        plt.savefig(f'{plot_outdir}/{fname}')
+        if verbose:
+            print(f'Figure {fname} stored in {plot_outdir}')
+        if show_plot:
+            plt.show()
+        plt.close()
+    
+    if make_plot:
+        fname = 'kgain_vs_mean'
+        plt.figure()
+        plt.plot(10**x_vals, kgain_arr, marker='o', linestyle='-', color='b', label='kgain')
+        # Set y-axis range
+        plt.ylim(6, 11)
+        # Add a horizontal line at y = 10
+        plt.axhline(y=8.7, color='r', linestyle='--', label='y = 8.7')
+        # Add a grid
+        plt.grid(True)
+        plt.title('kgain versus mean')
+        plt.savefig(f'{plot_outdir}/{fname}')
+        if verbose:
+            print(f'Figure {fname} stored in {plot_outdir}')
+        if show_plot:
+            plt.show()
+        plt.close()
+    
+    kgain_clipped, _ = sigma_clip(kgain_arr)
+    mode_kgain,_ = calculate_mode(kgain_clipped)
+
+    # adopt 'mode_kgain' as the final value to return
+    kgain = mode_kgain
+
+    if make_plot:
+        fname = 'kgain_clipped_histogram'
+        plt.figure()
+        # 'auto' lets matplotlib decide the number of bins
+        plt.hist(kgain_clipped, bins='auto', log=True)
+        plt.title('Histogram of clipped kgain values')
+        plt.savefig(f'{plot_outdir}/{fname}')
+        if verbose:
+            print(f'Figure {fname} stored in {plot_outdir}')
+        if show_plot:
+            plt.show()
+        plt.close()
+    
+    # parameter for plot
+    parm1 = -0.5*np.log10(kgain)
+    
+    # Gaussian read noise value in DN
+    mean_rn_gauss_DN = np.mean(rn_gauss)
```

Comment:
Is anything done with these? I feel like we discussed putting them into the ext_hdr as a keyword. Both the values and the errors.

Reply:
@maxwellmb: Read noise is a future task according to the latest convo on slack and the last telecon. https://github.com/roman-corgi/corgidrp/issues/135 

 Do you mean to include it now in this release? This implies changing KGain class accordingly. Is the read noise estimated elsewhere besides in calibrate_kgain? @kjl0025 I would like to avoid two different sources for the same type of noise, unless there's a good reason and some documentation on what to do with both estimations. If it is only in K-Gain, it's easier.

Reply:
Ok, good point - we shouldn't have two sources of read noise. My first instinct is that it should be where the PTC is defined, but I'm certainly open to other thoughts. @kjl0025 do you re-derive it yourself in your routine? 

It would be incredibly simple to just add these variables to the header at this point, isn't that what we talked about? I could be forgetting things.

Reply:
wrt the latter: my understanding from slack was yes and then no. I'm fine if it is yes again. The only caveat is what I raised above.

Reply:
as far as I can tell the read noise is not saved when creating the NoiseMaps

Reply:
if it is only a new ext_hdr keyword, you do not need to change the KGain class. just
kgain.ext_hdr["RNOISE"] = mean_rn_gauss_DN

Reply:
Read noise is calculated simply as a check in my noise maps code, but it isn't supposed to be the official calibration product for read noise, and as Max said, it isn't returned (in the DRP version of the code).  The way it is calculated in this k gain code is the official (more accurate) way to do it.  The official return for the read noise was called `mean_rn_gauss_e` originally (and perhaps still retains that name in this DRP version; I just haven't checked it lately).  Yeah, I thought we discussed putting that in a header.  Like Juergen said, we could do "RNOISE", and another header could be "RN_ERR".  Or to have them more similar and both <= 8 characters (to avoid HIERARCH cards), we could do maybe "RN" and "RN_ERR"?  So "RN" should be `mean_rn_gauss_e`, and the error could be an estimate of the standard deviation error (fractional error goes like 1/sqrt(2N)?  I'll have to check that later).

Reply:
Done. I added two keywords as suggested above. Build and tests pass.